File: 713-subarray-product-less-than-k/713-subarray-product-less-than-k.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var numSubarrayProductLessThanK = function(nums, k) {
    if (k <= 1) return 0;
    
    let [l, r, ans, prod] = [0, 0, 0, nums[0]];
    while (r < nums.length && l <= r) {
        // console.log(prod);
        if (prod < k) {
            ans += r - l + 1;
            r++;
            prod *= nums[r];
        } else if (l < r) {
            prod /= nums[l];
            l++;
        } else if (l === r) {
            r++;
            prod *= nums[r];
        }
    }
    return ans;
};

File: 713-subarray-product-less-than-k/713-subarray-product-less-than-k.rb
# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer}
def num_subarray_product_less_than_k(nums, k)
    l, r, ans, prod = 0, 0, 0, 1
    while r < nums.length
        prod *= nums[r]
        while prod >= k && l < r
            prod /= nums[l]
            l += 1
        end
        if prod < k
            ans += r - l + 1
        end
        r += 1
    end
    return ans
end